### PR TITLE
Loading circle google sign in

### PR DESCRIPTION
**Pull request by bryanporcello**
Addition of a loading progress circle when google is authenticating the user

### CODE REVIEW

File: app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt [hunk 1]
```diff
@@ -60,8 +66,8 @@ class SignInActivity : AppCompatActivity() {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_sign_in)
 
+        // Find the sign-in button and initialize firebaseAuth
         signInGoogleLayout =findViewById(R.id.signInGoogleLayout)
-
         firebaseAuth = FirebaseAuth.getInstance()
 
         // When the google sign in image is clicked, the google sign in page appears
```
Comment (iliasmerigh) [inline app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt:73]: Here also you should create a separate private helper function that handles the behavior of the UI component

File: app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt [hunk 2]
```diff
@@ -93,13 +99,43 @@ class SignInActivity : AppCompatActivity() {
...
-                        // Here I will hide the progress bar when implemented
+                        hideProgressDialog()
                         finish()
                     }
                 }
             }
     }
 
+    // Variable to hold the progress dialog
+    private var dialog: AlertDialog? = null
```
Comment (iliasmerigh) [inline app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt:110]: You can move up this variable to the top of the class so all variables are declared in the same place

File: app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt [hunk 3]
```diff
@@ -93,13 +99,43 @@ class SignInActivity : AppCompatActivity() {
...
+                window?.clearFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND)
+                show()
+            }
+    }
+
+
+    /**
+     * Hides the progress dialog.
+     */
+    private fun hideProgressDialog() {
```
Comment (iliasmerigh) [inline app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt:134]: You use two different approaches for the show and hide progress dialog. For better clarity, change this so its coherent with the _showProgressDialog_ :

```
if (dialog?.isShowing == false) {
            return
}


dialog!!.dismiss()

dialog = null```

File: app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt [hunk 4]
```diff
@@ -35,13 +40,14 @@ class SignInActivity : AppCompatActivity() {
         if (result.resultCode == Activity.RESULT_OK) {
```
Comment (iliasmerigh) [inline app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt:40]: When I tried to sign in, it didn't send me to the main activity as expected. After debugging a while I noticed that the problem comes from this condition. For some reason, the _result.resultCode_ is **CANCELLED** even if I (as a user) did everything right.

Below you can see the state of the debugger.

Try to fix this problem or tell me if I am the only one who has it

image

**Review by iliasmerigh — CHANGES_REQUESTED** [review]
Overall good PR that implements a difficult feature. However, the sign-in doesn't work as expected. Once this bug will be corrected, it will be a great improvement in the app !

Moreover, it could be good if you test some methods. The google sign-in is difficult to test (even impossible) so we don't expect you to test it. But I think that the RegisterForActivityResult could be tested as well as some UI components.

For testing the "RegisterForActivityResult" you can take inspration of a test I wrote in a previous version :

**In the class :**

```
cameraActivityLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            onCameraActivityResult(result)
        }
```
        
**In the test :**

```
// Define a correct result that the camera should return
        val cameraResult = Instrumentation.ActivityResult(Activity.RESULT_OK, correctCameraResponseIntent)

        // We expect that the camera activity will return "cameraResult"
        Intents.intending(IntentMatchers.hasAction(MediaStore.ACTION_IMAGE_CAPTURE)).respondWith(cameraResult)

        // Click the add new post button to start the camera Activity
        onView(withId(R.id.addNewPostBtn)).perform(ViewActions.click())

        // Verify that the camera activity was actually launched
        Intents.intended(IntentMatchers.hasAction(MediaStore.ACTION_IMAGE_CAPTURE))

        // Verify that the SharePostActivity has been launched
        Intents.intended(
            Matchers.allOf(
                IntentMatchers.hasComponent(SharePostActivity::class.java.name),
                IntentMatchers.hasExtra(CameraFragment.EXTRA_IMAGE_BITMAP, image)
            )
        )
```

**Comment by bryanporcello** [comment]
> Overall good PR that implements a difficult feature. However, the sign-in doesn't work as expected. Once this bug will be corrected, it will be a great improvement in the app !
> 
> Moreover, it could be good if you test some methods. The google sign-in is difficult to test (even impossible) so we don't expect you to test it. But I think that the RegisterForActivityResult could be tested as well as some UI components.
> 
> For testing the "RegisterForActivityResult" you can take inspration of a test I wrote in a previous version :
> 
> **In the class :**
> 
> ```
> cameraActivityLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
>             onCameraActivityResult(result)
>         }
> ```
> 
> **In the test :**
> 
> ```
> // Define a correct result that the camera should return
>         val cameraResult = Instrumentation.ActivityResult(Activity.RESULT_OK, correctCameraResponseIntent)
> 
>         // We expect that the camera activity will return "cameraResult"
>         Intents.intending(IntentMatchers.hasAction(MediaStore.ACTION_IMAGE_CAPTURE)).respondWith(cameraResult)
> 
>         // Click the add new post button to start the camera Activity
>         onView(withId(R.id.addNewPostBtn)).perform(ViewActions.click())
> 
>         // Verify that the camera activity was actually launched
>         Intents.intended(IntentMatchers.hasAction(MediaStore.ACTION_IMAGE_CAPTURE))
> 
>         // Verify that the SharePostActivity has been launched
>         Intents.intended(
>             Matchers.allOf(
>                 IntentMatchers.hasComponent(SharePostActivity::class.java.name),
>                 IntentMatchers.hasExtra(CameraFragment.EXTRA_IMAGE_BITMAP, image)
>             )
>         )
> ```

Hi Ilias, thanks a lot for your thoughtful review. Concerning the sign in, we already encounter and discuss the matter with Luca and David. We forgot to tell afterward, my apology. We realized that we need to insert your SHA1 in our Firebase account to use google sign in in developer mode. As you are not familiar with Firebase (yet?), you can simply give it to me and I will add it. 

To access your SHA1, go to your android terminal where you work on the project and type : ./gradlew signingReport   

After some time, your SHA1 will appear as in the photo below. Send it to me in private. 

I stay at your disposal may you need any help in retrieving your SHA1. 

After I insert it in Firebase, everything should be fine. 

image

File: app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt (hunk 2 above)
Comment (bryanporcello) [reply app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt:110]: Well spotted, thanks

File: app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt (hunk 3 above)
Comment (bryanporcello) [reply app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt:134]: Thanks for the proposition. However, I am not certain that this line below represents best Android practice : 

dialog!!.dismiss()

Those two methods are in fact quite different therefore I don't think that is very useful to increase the number of line codes of hideProgressDialog() for the sole purpose to be more coherent with ShowProgressDialog. 

If my explanation is ok for you, I'd prefer to leave it as it is. What do you think ?

File: app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt (hunk 4 above)
Comment (bryanporcello) [reply app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt:40]: See my comments below about SHA 1. It should resolve the problem. Let me know if you have any questions.

[the remark this answers is not in the document]
Comment (bryanporcello) [reply app/src/main/java/com/github/sdp_begreen/begreen/activities/SignInActivity.kt:43]: You are right,  thanks. It's easier to read the code like that.

**Review by iliasmerigh — APPROVED** [review]
After discussing with Bryan, we have come to the realization that the test we proposed in the pull request is not as feasible as we initially thought. Unfortunately, we were unable to successfully test it. As a result, we will need to revisit this matter at a later time.

Additionally, we would like to note that the sign-in feature is a part of the tests that we are unable to cover due to its difficulty.

**Pull request merged**